mods/parsoid.js

```diff
@@ -150,54 +189,72 @@ PSP.getRevisionInfo = function(restbase, req) {
-                uri: self.getBucketURI(rp, format, rp.tid)
-            };
-            var contentReq = restbase.get(beReq)
-            .catch(function(e) {
-                if (e.status === 404) {
-                    return self.getRevisionInfo(restbase, req)
-                    .then(function(revInfo) {
-                        rp.revision = revInfo.rev + '';
-                        if (revInfo.title !== rp.title) {
-                            // Re-try to retrieve from storage with the
-                            // normalized title & revision
-                            rp.title = revInfo.title;
-                            return _getFormat(restbase, req);
-                        } else {
-                            return self.generateAndSave(restbase, req, format, uuid.v1());
-                        }
-                    });
+    var rp = req.params;
+    rp.title = normalizeTitle(rp.title);
+
+    function generateContent (storageRes) {
+        if (storageRes.status === 404 || storageRes.status === 200) {
+            return self.getRevisionInfo(restbase, req)
+            .then(function(revInfo) {
+                rp.revision = revInfo.rev + '';
+                if (revInfo.title !== rp.title) {
+                    // Re-try to retrieve from storage with the
+                    // normalized title & revision
+                    rp.title = revInfo.title;
+                    return self.getFormat(format, restbase, req);
                 } else {
-                    // Don't generate content if there's some other error.
-                    throw e;
+                    return self.generateAndSave(restbase, req, format, storageRes);
                 }
             });
-            return self.wrapContentReq(restbase, req, contentReq);
+        } else {
+            // Don't generate content if there's some other error.
+            throw storageRes;
         }
-    };
+    }
+
+    var contentReq = restbase.get({
+        uri: self.getBucketURI(rp, format, rp.tid)
+    });
+
+    if (req.headers && /no-cache/i.test(req.headers['cache-control'])
+            && rp.revision)
+    {
+        // Check content generation either way
+        return contentReq.then(function(res) {
+                if (req.headers['if-unmodified-since']) {
+                    try {
+                        var jobTime = new Date(req.headers['if-unmodified-since']);
+                        if (uuid.v1time(res.headers.etag) >= jobTime) {
+                            // Already up to date, nothing to do.
+                            return {
+                                status: 412,
+                                body: {
+                                    type: 'precondition_failed',
+                                    detail: 'The precondition failed'
+                                }
+                            };
+                        }
+                    } catch (e) { console.log(e); } // Ignore errors from date parsing
```

Comment:
If we want to stick to the [spec](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html), we should actually ignore `If-Unmodified-Since` when the date is invalid.